specs/CDPSessionIdSupport.md

```diff
@@ -0,0 +1,328 @@
+                std::wstring targetId = GetJSONStringField(jsonMessage.get(), L"targetId");
+                if (m_devToolsTargetDescriptionMap.find(targetId) !=
+                    m_devToolsTargetDescriptionMap.end())
+                {
+                    // This is a target that we are interested in, update description.
+                    std::wstring type = GetJSONStringField(jsonMessage.get(), L"type");
+                    std::wstring url = GetJSONStringField(jsonMessage.get(), L"url");
+                    m_devToolsTargetDescriptionMap[targetId] = type + L"," + url;
+                }
+                return S_OK;
+            })
+            .Get(),
+        &m_targetInfoChangedToken));
+    // Enable Runtime events for the default target of top page to receive Runtime.consoleAPICalled events, which is fired when console.log is called.
+    m_webView->CallDevToolsProtocolMethod(L"Runtime.enable", L"{}", nullptr);
+    // Auto attach to iframe and dedicated worker targets created from the default target of top page.
+    m_webView->CallDevToolsProtocolMethod(
+        L"Target.setAutoAttach",
+        LR"({"autoAttach":true,"waitForDebuggerOnStart":false,"flatten":true})", nullptr);
+}
+```
+## WinRT and .NET    
+```c#
+// The sessions and targets descriptions are tracked by 2 dictionaries:
+// SessionId to TargetId dictionary: m_devToolsSessionMap;
+// TargetId to description dictionary, where description is "<target type>,<target url>": m_devToolsTargetInfoMap
+// GetJSONStringField is a helper function that can retrieve a string field from a json message.
+
+private void CoreWebView2_ConsoleAPICalled(CoreWebView2 sender, CoreWebView2DevToolsProtocolEventReceivedEventArgs args)
+{
+   // Figure out which target the console.log comes from
+   string eventSource;
+   string sessionId = args.SessionId;
+   if (sessionId.Length > 0)
+   {
+     string targetId = m_devToolsSessionMap[sessionId];
+     eventSource = m_devToolsTargetDescriptionMap[targetId];
+   }
+  // App code to log these events.
+  LogConsoleLogMessage(eventSource, args.ParameterObjectAsJson);
+}
+
+private void CoreWebView2_AttachedToTarget(CoreWebView2 sender, CoreWebView2DevToolsProtocolEventReceivedEventArgs args)
+{
+  // A new target is attached, add its info to maps.
+  string jsonMessage = args.ParameterObjectAsJson;
+  string sessionId = GetJSONStringField(jsonMessage, L"sessionId");
+  string targetId = GetJSONStringField(jsonMessage, L"targetId");
+  m_devToolsSessionMap[sessionId] = targetId;
+  string type = GetJSONStringField(jsonMessage, L"type");
+  string url = GetJSONStringField(jsonMessage, L"url");
+  m_devToolsTargetDescriptionMap[targetId] = type + L"," + url;
+  // Auto attach to targets further created from this target.
+  _ = m_webview.CallDevToolsProtocolMethodAsync("Target.setAutoAttach",
+        @"{""autoAttach"":true,""waitForDebuggerOnStart"":false,""flatten"":true}",
+        sessionId);
```

Comment:
This gets tricky if we ever need to add a third parameter to CallDevToolsProtocolMethodAsync, say, `string telemetryId`. We already have

1. CallDevToolsProtocolMethodAsync(string methodName, string parametersAsJson)
2. CallDevToolsProtocolMethodAsync(string methodName, string parametersAsJson, string sessionId)

now where do we put the telemetryId?

We can't put it in the third slot:
* CallDevToolsProtocolMethodAsync(string methodName, string parametersAsJson, string telemetryId) 🡄 conflict with 2
* CallDevToolsProtocolMethodAsync(string methodName, string parametersAsJson, string telemetryId, string sessionId)

We would have to put it in the fourth slot:
* CallDevToolsProtocolMethodAsync(string methodName, string parametersAsJson, string sessionId, string telemetryId)

I guess that's okay because people who want to operate on the top document can pass `string.Empty` as the `sessionId`.

The `ForSession` pattern used by C++ parallels the Windows Runtime `ForUser` pattern, using a separate method name so we don't create future overload conflicts. But maybe this feature area is constrained enough that we don't anticipate overload conflicts being a problem.

Reply:
Do we prefer using a different API name CallDevToolsProtocolMethodForSessionAsync instead of overload CallDevToolsProtocolMethodAsync? I personally prefer using a different name.

Reply:
Discussion has no one is particularly excited for either. If we do ForSession we can put the session parameter first, where it makes more sense anyway.

Change to `CallDevToolsProtocolMethodForSessionAsync(string sessionId, ...)` instead of overload